storage/src/snapshot_cache.rs

```diff
@@ -0,0 +1,268 @@
+///
+/// The first `Snapshot` inserted is allowed to have a "pre-genesis" version number
+/// of [`u64::MAX`], but every subsequent inserts must have consecutive versions.
+pub struct SnapshotCache {
+    /// A sequence of increasingly recent [`Snapshot`]s.
+    cache: VecDeque<Snapshot>,
+    /// The max length and capacity of the cache.
+    max_size: usize,
+}
+
+impl SnapshotCache {
+    /// Creates a [`SnapshotCache`] with `max_size` capacity, and inserts an initial `Snapshot` in
+    /// it.
+    ///
+    /// ## Panics
+    ///
+    /// This function panics if the specified capacity is zero.
+    pub fn new(initial: Snapshot, max_size: usize) -> Self {
+        if max_size == 0 {
+            panic!("snapshot_cache: capacity must be greater than 0.");
+        }
+        let mut cache = VecDeque::with_capacity(max_size);
+        cache.push_front(initial);
+
+        Self { cache, max_size }
+    }
+
+    /// Attempts to insert a [`Snapshot`] entry into the cache. If the cache is full, the oldest
+    /// entry will be evicted to make space.
+    ///
+    /// [`Snapshot`]s must be inserted sequentially relative to their `jmt::Version`s and have
+    /// consecutive version numbers.
+    ///
+    /// ## Special case
+    ///
+    /// The first [`Snapshot`] inserted is allowed to have version `u64::MAX` to support the
+    /// pre-genesis case.
+    ///
+    /// ## Errors
+    ///
+    /// The method will return an error if the supplied `snapshot` has a version number that is:
+    ///
+    /// - stale i.e. older than the latest snapshots
+    ///
+    /// - skipping a version i.e. the difference between their version numbers is greater than 1
+    pub fn try_push(&mut self, snapshot: Snapshot) -> anyhow::Result<()> {
+        let latest_version = self.latest().version();
+        if latest_version.wrapping_add(1) != snapshot.version() {
+            return Err(anyhow!("snapshot_cache: trying to insert stale snapshots."));
+        }
+
+        if self.cache.len() >= self.max_size {
+            self.cache.pop_back();
+        }
+
+        self.cache.push_front(snapshot);
+        Ok(())
+    }
+
+    /// Inserts a [`Snapshot`] entry in the cache. If the cache is full, then the oldest entry will
+    /// be evicted to make space.
+    ///
+    /// [`Snapshot`]s must be inserted sequentially relative to their `jmt::Version`s and have
+    /// consecutive version numbers.
+    ///
+    /// ## Special case
+    ///
+    /// The first [`Snapshot`] inserted is allowed to have version `u64::MAX` to support the
+    /// pre-genesis case.
+    ///
+    /// ## Panics
+    ///
+    /// This function panics if the supplied snapshot's version is:
+    ///
+    /// - stale i.e. older than the most recently inserted snapshot
+    ///
+    /// - skipping a version i.e. the difference between the version numbers is greater than 1.
+    ///
+    /// For a non-faillible version, see [`try_push`].
+    pub fn push(&mut self, snapshot: Snapshot) {
```

Comment:
Since this is just wrapping `try_push` with an `expect`, I think we could delete it and just have the caller use `expect` if they don't want to deal with an error.